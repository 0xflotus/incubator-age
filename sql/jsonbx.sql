--
-- JSONBX data type regression tests
--

--
-- Load extension and set path
--
LOAD 'agensgraph';
SET search_path TO ag_catalog;

--
-- Create a table using the JSONBX type
--
CREATE TABLE jsonbx_table (type text, jsonbx jsonbx);

--
-- Insert values to exercise jsonbx_in/jsonbx_out
--
INSERT INTO jsonbx_table VALUES ('bool', 'true');
INSERT INTO jsonbx_table VALUES ('bool', 'false');

INSERT INTO jsonbx_table VALUES ('null', 'null');

INSERT INTO jsonbx_table VALUES ('string', '""');
INSERT INTO jsonbx_table VALUES ('string', '"This is a string"');

INSERT INTO jsonbx_table VALUES ('integer8', '0');
INSERT INTO jsonbx_table VALUES ('integer8', '9223372036854775807');
INSERT INTO jsonbx_table VALUES ('integer8', '-9223372036854775808');

INSERT INTO jsonbx_table VALUES ('float8', '0.0');
INSERT INTO jsonbx_table VALUES ('float8', '1.0');
INSERT INTO jsonbx_table VALUES ('float8', '-1.0');
INSERT INTO jsonbx_table VALUES ('float8', '100000000.000001');
INSERT INTO jsonbx_table VALUES ('float8', '-100000000.000001');
INSERT INTO jsonbx_table VALUES ('float8', '0.00000000000000012345');
INSERT INTO jsonbx_table VALUES ('float8', '-0.00000000000000012345');

INSERT INTO jsonbx_table VALUES ('integer8 array',
	'[-9223372036854775808, -1, 0, 1, 9223372036854775807]');
INSERT INTO jsonbx_table VALUES('float8 array',
	'[-0.00000000000000012345, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 0.00000000000000012345]');
INSERT INTO jsonbx_table VALUES('mixed array', '[true, false, null, "string", 1, 1.0, {"bool":true}]');

INSERT INTO jsonbx_table VALUES('object', '{"bool":true, "null":null, "string":"string", "integer8":1, "float8":1.2, "arrayi8":[-1,0,1], "arrayf8":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int8":1, "float8":8.0}}');

SELECT * FROM jsonbx_table;

--
-- These should fail
--
INSERT INTO jsonbx_table VALUES ('bad integer8', '9223372036854775808');
INSERT INTO jsonbx_table VALUES ('bad integer8', '-9223372036854775809');
INSERT INTO jsonbx_table VALUES ('bad float8', 'NaN');
INSERT INTO jsonbx_table VALUES ('bad float8', 'Infinity');
INSERT INTO jsonbx_table VALUES ('bad float8', '-Infinity');

--
-- Cleanup
--
DROP TABLE jsonbx_table;

--
-- End of JSONBX data type regression tests
--
