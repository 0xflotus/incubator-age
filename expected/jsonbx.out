--
-- AGTYPE data type regression tests
--
--
-- Load extension and set path
--
LOAD 'agensgraph';
SET search_path TO ag_catalog;
--
-- Create a table using the AGTYPE type
--
CREATE TABLE agtype_table (type text, agtype agtype);
--
-- Insert values to exercise agtype_in/agtype_out
--
INSERT INTO agtype_table VALUES ('bool', 'true');
INSERT INTO agtype_table VALUES ('bool', 'false');
INSERT INTO agtype_table VALUES ('null', 'null');
INSERT INTO agtype_table VALUES ('string', '""');
INSERT INTO agtype_table VALUES ('string', '"This is a string"');
INSERT INTO agtype_table VALUES ('integer', '0');
INSERT INTO agtype_table VALUES ('integer', '9223372036854775807');
INSERT INTO agtype_table VALUES ('integer', '-9223372036854775808');
INSERT INTO agtype_table VALUES ('float', '0.0');
INSERT INTO agtype_table VALUES ('float', '1.0');
INSERT INTO agtype_table VALUES ('float', '-1.0');
INSERT INTO agtype_table VALUES ('float', '100000000.000001');
INSERT INTO agtype_table VALUES ('float', '-100000000.000001');
INSERT INTO agtype_table VALUES ('float', '0.00000000000000012345');
INSERT INTO agtype_table VALUES ('float', '-0.00000000000000012345');
INSERT INTO agtype_table VALUES ('integer array',
	'[-9223372036854775808, -1, 0, 1, 9223372036854775807]');
INSERT INTO agtype_table VALUES('float array',
	'[-0.00000000000000012345, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 0.00000000000000012345]');
INSERT INTO agtype_table VALUES('mixed array', '[true, false, null, "string", 1, 1.0, {"bool":true}]');
INSERT INTO agtype_table VALUES('object', '{"bool":true, "null":null, "string":"string", "integer":1, "float":1.2, "arrayi":[-1,0,1], "arrayf":[-1.0, 0.0, 1.0], "object":{"bool":true, "null":null, "string":"string", "int":1, "float":8.0}}');
SELECT * FROM agtype_table;
     type      |                                                                                                       agtype                                                                                                       
---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 bool          | true
 bool          | false
 null          | null
 string        | ""
 string        | "This is a string"
 integer       | 0
 integer       | 9223372036854775807
 integer       | -9223372036854775808
 float         | 0.0
 float         | 1.0
 float         | -1.0
 float         | 100000000.000001
 float         | -100000000.000001
 float         | 1.2345e-16
 float         | -1.2345e-16
 integer array | [-9223372036854775808, -1, 0, 1, 9223372036854775807]
 float array   | [-1.2345e-16, -100000000.000001, -1.0, 0.0, 1.0, 100000000.000001, 1.2345e-16]
 mixed array   | [true, false, null, "string", 1, 1.0, {"bool": true}]
 object        | {"bool": true, "null": null, "float": 1.2, "arrayf": [-1.0, 0.0, 1.0], "arrayi": [-1, 0, 1], "object": {"int": 1, "bool": true, "null": null, "float": 8.0, "string": "string"}, "string": "string", "integer": 1}
(19 rows)

--
-- These should fail
--
INSERT INTO agtype_table VALUES ('bad integer', '9223372036854775808');
ERROR:  value "9223372036854775808" is out of range for type bigint
LINE 1: INSERT INTO agtype_table VALUES ('bad integer', '92233720368...
                                                        ^
INSERT INTO agtype_table VALUES ('bad integer', '-9223372036854775809');
ERROR:  value "-9223372036854775809" is out of range for type bigint
LINE 1: INSERT INTO agtype_table VALUES ('bad integer', '-9223372036...
                                                        ^
INSERT INTO agtype_table VALUES ('bad float', 'NaN');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', 'NaN');
                                                      ^
DETAIL:  Token "NaN" is invalid.
CONTEXT:  AGTYPE data, line 1: NaN
INSERT INTO agtype_table VALUES ('bad float', 'Infinity');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', 'Infinity');
                                                      ^
DETAIL:  Token "Infinity" is invalid.
CONTEXT:  AGTYPE data, line 1: Infinity
INSERT INTO agtype_table VALUES ('bad float', '-Infinity');
ERROR:  invalid input syntax for type agtype
LINE 1: INSERT INTO agtype_table VALUES ('bad float', '-Infinity');
                                                      ^
DETAIL:  Token "-Infinity" is invalid.
CONTEXT:  AGTYPE data, line 1: -Infinity
--
-- Cleanup
--
DROP TABLE agtype_table;
--
-- End of AGTYPE data type regression tests
--
